File: backend/app/api/v1/admin/endpoints/observability.py
"""
Admin observability endpoints.
"""

from __future__ import annotations

from typing import Any
from uuid import UUID

from fastapi import APIRouter, Depends, Query

from app.api.deps import PermissionChecker
from app.models.user import User
from app.schemas.response import Response, success
from app.services import admin_observability

router = APIRouter()


def _cache_params(**params: Any) -> dict[str, Any]:
    return {key: value for key, value in params.items() if value is not None}


@router.get("/overview", response_model=Response[dict])
async def get_observability_overview(
    time_range: str = Query("30d", description="Time range: 7d, 30d, 90d, all"),
    current_user: User = Depends(PermissionChecker("admin:dashboard:access")),
) -> Any:
    data = await admin_observability.cached_payload(
        "overview",
        _cache_params(time_range=time_range),
        lambda: admin_observability.get_overview(time_range),
    )
    return success(data=data)


@router.get("/agents", response_model=Response[dict])
async def get_observability_agents(
    time_range: str = Query("30d", description="Time range: 7d, 30d, 90d, all"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    sort_by: str = Query("requests"),
    sort_order: str = Query("desc", pattern="^(asc|desc)$"),
    current_user: User = Depends(PermissionChecker("admin:dashboard:access")),
) -> Any:
    data = await admin_observability.cached_payload(
        "agents",
        _cache_params(
            time_range=time_range,
            page=page,
            page_size=page_size,
            sort_by=sort_by,
            sort_order=sort_order,
        ),
        lambda: admin_observability.get_agents(
            time_range, page, page_size, sort_by, sort_order
        ),
    )
    return success(data=data)


@router.get("/agent/{agent_id}", response_model=Response[dict])
async def get_observability_agent_detail(
    agent_id: UUID,
    time_range: str = Query("30d", description="Time range: 7d, 30d, 90d, all"),
    current_user: User = Depends(PermissionChecker("admin:dashboard:access")),
) -> Any:
    data = await admin_observability.cached_payload(
        "agent-detail",
        _cache_params(agent_id=str(agent_id), time_range=time_range),
        lambda: admin_observability.get_agent_detail(agent_id, time_range),
    )
    return success(data=data)


@router.get("/workflows", response_model=Response[dict])
async def get_observability_workflows(
    time_range: str = Query("30d", description="Time range: 7d, 30d, 90d, all"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    sort_by: str = Query("runs"),
    sort_order: str = Query("desc", pattern="^(asc|desc)$"),
    current_user: User = Depends(PermissionChecker("admin:dashboard:access")),
) -> Any:
    data = await admin_observability.cached_payload(
        "workflows",
        _cache_params(
            time_range=time_range,
            page=page,
            page_size=page_size,
            sort_by=sort_by,
            sort_order=sort_order,
        ),
        lambda: admin_observability.get_workflows(
            time_range, page, page_size, sort_by, sort_order
        ),
    )
    return success(data=data)


@router.get("/workflow/{workflow_id}", response_model=Response[dict])
async def get_observability_workflow_detail(
    workflow_id: UUID,
    time_range: str = Query("30d", description="Time range: 7d, 30d, 90d, all"),
    current_user: User = Depends(PermissionChecker("admin:dashboard:access")),
) -> Any:
    data = await admin_observability.cached_payload(
        "workflow-detail",
        _cache_params(workflow_id=str(workflow_id), time_range=time_range),
        lambda: admin_observability.get_workflow_detail(workflow_id, time_range),
    )
    return success(data=data)


@router.get("/timeouts", response_model=Response[dict])
async def get_observability_timeouts(
    time_range: str = Query("30d", description="Time range: 7d, 30d, 90d, all"),
    source: str = Query("all", pattern="^(all|agent|workflow)$"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    current_user: User = Depends(PermissionChecker("admin:dashboard:access")),
) -> Any:
    data = await admin_observability.cached_payload(
        "timeouts",
        _cache_params(
            time_range=time_range, source=source, page=page, page_size=page_size
        ),
        lambda: admin_observability.get_timeouts(time_range, source, page, page_size),
    )
    return success(data=data)


@router.get("/throughput", response_model=Response[dict])
async def get_observability_throughput(
    time_range: str = Query("30d", description="Time range: 7d, 30d, 90d, all"),
    granularity: str | None = Query(None, description="Granularity: hour, day"),
    current_user: User = Depends(PermissionChecker("admin:dashboard:access")),
) -> Any:
    data = await admin_observability.cached_payload(
        "throughput",
        _cache_params(time_range=time_range, granularity=granularity),
        lambda: admin_observability.get_throughput(time_range, granularity),
    )
    return success(data=data)


@router.get("/tokens", response_model=Response[dict])
async def get_observability_tokens(
    time_range: str = Query("30d", description="Time range: 7d, 30d, 90d, all"),
    current_user: User = Depends(PermissionChecker("admin:dashboard:access")),
) -> Any:
    data = await admin_observability.cached_payload(
        "tokens",
        _cache_params(time_range=time_range),
        lambda: admin_observability.get_tokens(time_range),
    )
    return success(data=data)


@router.get("/system/health", response_model=Response[dict])
async def get_observability_system_health(
    current_user: User = Depends(PermissionChecker("admin:dashboard:access")),
) -> Any:
    data = await admin_observability.cached_payload(
        "system-health",
        {},
        admin_observability.get_system_health,
    )
    return success(data=data)


@router.get("/system/trend", response_model=Response[dict])
async def get_observability_system_trend(
    current_user: User = Depends(PermissionChecker("admin:dashboard:access")),
) -> Any:
    data = await admin_observability.cached_payload(
        "system-trend",
        {},
        admin_observability.get_system_trend,
    )
    return success(data=data)


@router.get("/system/slow-queries", response_model=Response[dict])
async def get_observability_slow_queries(
    threshold_ms: int = Query(1000, ge=1),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    current_user: User = Depends(PermissionChecker("admin:dashboard:access")),
) -> Any:
    data = await admin_observability.cached_payload(
        "system-slow-queries",
        _cache_params(threshold_ms=threshold_ms, page=page, page_size=page_size),
        lambda: admin_observability.get_slow_queries(threshold_ms, page, page_size),
    )
    return success(data=data)


@router.get("/system/workers", response_model=Response[dict])
async def get_observability_workers(
    current_user: User = Depends(PermissionChecker("admin:dashboard:access")),
) -> Any:
    data = await admin_observability.cached_payload(
        "system-workers",
        {},
        admin_observability.get_workers,
    )
    return success(data=data)

File: frontend/lib/api/admin/observability.ts
import { api } from '../client'

export type ObservabilityTimeRange = '7d' | '30d' | '90d' | 'all'
export type ObservabilitySortOrder = 'asc' | 'desc'

export interface ObservabilityPercentiles {
  p50_ms: number | null
  p90_ms: number | null
  p95_ms: number | null
  p99_ms?: number | null
}

export interface ObservabilityOverview {
  time_range: ObservabilityTimeRange
  generated_at: string
  cache_ttl_seconds: number
  totals: {
    agent_requests: number
    workflow_runs: number
    total_requests: number
    total_tokens: number
  }
  rates: {
    agent_success_rate: number
    workflow_success_rate: number
    overall_success_rate: number
    timeout_rate: number
  }
  latency: ObservabilityPercentiles
  ttft?: ObservabilityPercentiles
  throughput: {
    current_qps: number
    peak_hourly_requests: number
  }
}

export interface ObservabilityPage<T> {
  items: T[]
  total: number
  page: number
  page_size: number
  time_range?: ObservabilityTimeRange
}

export interface AgentPerformanceRow extends ObservabilityPercentiles {
  agent_id: string
  agent_name: string | null
  team_name: string | null
  request_count: number
  success_count: number
  error_count: number
  timeout_count: number
  total_tokens: number
  ttft_p50_ms?: number | null
  ttft_p90_ms?: number | null
  ttft_p95_ms?: number | null
  ttft_p99_ms?: number | null
  success_rate: number
  timeout_rate: number
  avg_tokens: number
}

export interface WorkflowPerformanceRow extends ObservabilityPercentiles {
  workflow_id: string
  workflow_name: string | null
  team_name: string | null
  run_count: number
  success_count: number
  error_count: number
  timeout_count: number
  failed_nodes: number
  avg_nodes: number | null
  total_tokens: number
  success_rate: number
  timeout_rate: number
  avg_tokens: number
}

export interface ObservabilityTrendPoint {
  bucket: string
  request_count?: number
  run_count?: number
  p50_ms?: number | null
  p90_ms?: number | null
  p95_ms?: number | null
  ttft_p95_ms?: number | null
  success_rate?: number
  timeout_count?: number
  failed_nodes?: number
}

export interface AgentDetailResponse {
  time_range: ObservabilityTimeRange
  agent: AgentPerformanceRow | null
  trend: ObservabilityTrendPoint[]
}

export interface WorkflowDetailResponse {
  time_range: ObservabilityTimeRange
  workflow: WorkflowPerformanceRow | null
  trend: ObservabilityTrendPoint[]
  nodes: Array<{
    node_type: string
    execution_count: number
    failed_count: number
    avg_duration_ms: number | null
  }>
}

export interface TimeoutEvent {
  source: 'agent' | 'workflow'
  entity_id: string | null
  entity_name: string
  model: string | null
  timeout_type: string
  created_at: string | null
  duration_ms: number | null
  status: string | null
}

export interface TimeoutResponse extends ObservabilityPage<TimeoutEvent> {
  distribution: Record<string, number>
  agent_timeout_type_available: boolean
  note: string
}

export interface ThroughputResponse {
  time_range: ObservabilityTimeRange
  granularity: string
  current: {
    qps: number
    tps: number
    running_workflows: number
  }
  buckets: Array<{
    bucket: string
    agent_requests: number
    workflow_runs: number
    total_requests: number
  }>
}

export interface TokenResponse {
  time_range: ObservabilityTimeRange
  total_tokens: number
  by_source: Array<{ source: string; tokens: number }>
  by_model: Array<{ model: string; tokens: number }>
}

export interface SystemHealthResponse {
  generated_at: string
  cache_ttl_seconds: number
  cpu: Record<string, unknown>
  memory: Record<string, unknown>
  disk: Record<string, unknown>
  database: Record<string, unknown>
  redis: Record<string, unknown>
  workers: WorkerResponse
}

export interface SystemTrendResponse {
  items: Array<Record<string, unknown>>
}

export interface SlowQueriesResponse {
  available: boolean
  reason: string | null
  items: Array<Record<string, unknown>>
  total: number
  page: number
  page_size: number
}

export interface WorkerResponse {
  status: string
  worker_count: number
  active_tasks: number
  reserved_tasks: number
  scheduled_tasks: number
  queues: Array<{ queue: string; pending: number }>
  error?: string
}

function withParams(path: string, params: Record<string, string | number | undefined>) {
  const query = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined) {
      query.set(key, String(value))
    }
  })
  const serialized = query.toString()
  return serialized ? `${path}?${serialized}` : path
}

export const observabilityApi = {
  getOverview: (timeRange: ObservabilityTimeRange = '30d'): Promise<ObservabilityOverview> =>
    api.get<ObservabilityOverview>(withParams('/admin/observability/overview', { time_range: timeRange })),

  getAgents: (params: {
    time_range?: ObservabilityTimeRange
    page?: number
    page_size?: number
    sort_by?: string
    sort_order?: ObservabilitySortOrder
  } = {}): Promise<ObservabilityPage<AgentPerformanceRow>> =>
    api.get<ObservabilityPage<AgentPerformanceRow>>(withParams('/admin/observability/agents', params)),

  getAgentDetail: (agentId: string, timeRange: ObservabilityTimeRange = '30d'): Promise<AgentDetailResponse> =>
    api.get<AgentDetailResponse>(withParams(`/admin/observability/agent/${agentId}`, { time_range: timeRange })),

  getWorkflows: (params: {
    time_range?: ObservabilityTimeRange
    page?: number
    page_size?: number
    sort_by?: string
    sort_order?: ObservabilitySortOrder
  } = {}): Promise<ObservabilityPage<WorkflowPerformanceRow>> =>
    api.get<ObservabilityPage<WorkflowPerformanceRow>>(withParams('/admin/observability/workflows', params)),

  getWorkflowDetail: (workflowId: string, timeRange: ObservabilityTimeRange = '30d'): Promise<WorkflowDetailResponse> =>
    api.get<WorkflowDetailResponse>(withParams(`/admin/observability/workflow/${workflowId}`, { time_range: timeRange })),

  getTimeouts: (params: {
    time_range?: ObservabilityTimeRange
    source?: 'all' | 'agent' | 'workflow'
    page?: number
    page_size?: number
  } = {}): Promise<TimeoutResponse> =>
    api.get<TimeoutResponse>(withParams('/admin/observability/timeouts', params)),

  getThroughput: (params: {
    time_range?: ObservabilityTimeRange
    granularity?: 'hour' | 'day'
  } = {}): Promise<ThroughputResponse> =>
    api.get<ThroughputResponse>(withParams('/admin/observability/throughput', params)),

  getTokens: (timeRange: ObservabilityTimeRange = '30d'): Promise<TokenResponse> =>
    api.get<TokenResponse>(withParams('/admin/observability/tokens', { time_range: timeRange })),

  getSystemHealth: (): Promise<SystemHealthResponse> =>
    api.get<SystemHealthResponse>('/admin/observability/system/health'),

  getSystemTrend: (): Promise<SystemTrendResponse> =>
    api.get<SystemTrendResponse>('/admin/observability/system/trend'),

  getSlowQueries: (params: { threshold_ms?: number; page?: number; page_size?: number } = {}): Promise<SlowQueriesResponse> =>
    api.get<SlowQueriesResponse>(withParams('/admin/observability/system/slow-queries', params)),

  getWorkers: (): Promise<WorkerResponse> =>
    api.get<WorkerResponse>('/admin/observability/system/workers'),
}
